### ISSUE TITLE
Issue with inserting image with transparency into PF

### ISSUE THREAD

**Issue by iWHOswald** [pymupdf/PyMuPDF#559]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
I have found a bug while using the fitz module to insert images into a PDF in V. 1.17.2. I think it is specific to an image with transparency, as when I hard code inserting an image with no transparency it works fine. However, attempting to insert one with transparency results in no image showing in the saved PDF and adobe reader prompting the message stating an error was found in the PDF. I found that reverting to V. 1.16.18 does not have this bug, so it must be related to the new versions.

## To Reproduce (mandatory)
The following snippet of code resulted in the erroneous PDF:

`                x_terp_first = 29
                y_terp = 300
                scaler = 48
                for i in range(5):
                    if i == 0:
                        x_terp1 = x_terp_first
                    else:
                        x_terp1 = x_terp1 + 45
                    img2 = fitz.Rect(x_terp1, y_terp, x_terp1 + 24, y_terp + scaler)  # where we want to put the image
                    pix_terp = fitz.Pixmap(terp_plots[i])  # any supported image file
                    page.insertImage(img2, pixmap=pix_terp, overlay=True)  # insert image
`
terp_plots[i] is an array of 5 image files. When this is replaced with an image that has no transparency, it works. With transparency results in the error.

## Expected behavior (optional)
Ideally want image to appear in PDF; as said above, V. 1.16.18 fixes this issue.

## Your configuration (mandatory)
 - Windows 10 64-bit
 - Python 3.8.2 64-bit
 - PyMuPDF version 1.17.2, pip installed

**Comment by JorjMcKie** [comment]
cannot reproduce
here is a zip containing 2 scripts and a test PDF:
* `pixmapper.py` reads the test PDF and makes a transparent image of each of the 4 pages
* `inserter.py` makes a new PDF and inserts the 4 page PNGs into respective rectangles of its one and only page

The resulting new PDF looks good and can be read by the usual viewers including Adobe.
[transparency.zip](https://github.com/pymupdf/PyMuPDF/files/4911363/transparency.zip)

**Comment by JorjMcKie** [comment]
Can you please let me have the problem images for reproducing the error?

**Comment by iWHOswald** [comment]
Hi Jorj: 

Yes, please see attached. The files are ordered in the array as such:

[white100_100.png, white100_44.png, white100_5.png, white100_5.png, white100_3.png]

The first image which has no transparency has no issue pasting, but the remaining that do do not work. I don't actually know if it's the transparency that is the problem, it's just the only thing I could think of. The fact it works in 1.16.18 and not 1.17.2 also is odd.

[figs.zip](https://github.com/pymupdf/PyMuPDF/files/4919987/figs.zip)

**Comment by JorjMcKie** [comment]
Issue confirmed.
It does have to do with both, transparency and MuPDF version:
Transparent images are stored in PDF with two image objects: (1) the image itself and (2) a "Mask" image containing the transparency info for each pixel of the image.
Your examples result in masks consisting of 1-bit per image pixel, because the images in the end are gray level - allthough technically provided as RGB.
MuPDF v1.17.0 announced these changes among others:
```
...

- pdf: Redaction now works on images and links as well as text.
- pdf: Greek, Cyrillic, Chinese, Japanese, and Korean scripts in forms and annotations.
- pdf: File attachment annotations.
- pdf: Use CCITT Fax compression for 1-bit images when creating PDF files.
...
```
The last bullet (CCITT Fax compression) is where something goes wrong: MuPDF compresses the mask images with that new compression algorithm, which must have flaws.
If you save the created PDF using `deflate=False` or ``expnd=255``, it **_can_** be displayed with Adobe. Of course the PDF is quite large.
If you later try to compress the PDF, e.g. via ``mutool clean -z ...``, an erroneous PDF again results.
If you agree, I will send the problem to MuPDF's bug system.

**Comment by iWHOswald** [comment]
Excellent diagnosis! This must be the culprit. Yes, I think passing this info along to the MuPDF folks could be useful. Thanks again for the help!

**Comment by JorjMcKie** [comment]
To be honest, I have already submitted a minimal reproducer to them 🙄, this is the link: https://bugs.ghostscript.com/show_bug.cgi?id=702564.
You may want to underpin with our own assessment - always a good idea to broaden the audience ...

**Comment by JorjMcKie** [comment]
If you are looking for way to insert images without having to bear big files, you may to consider this approach:

Right before saving the document, I scanned the page's images and compressed them one by one. This prevents MuPDF's saving mechanism to recompress anything already well compressed.
For those 1-bit mask images, this means they get compressed the old way (deflate / gzip), and MuPDF then won't touch them.

```python
>>> for img in page.getImageList():  # scan through all the page's images
	x0 = img[0]  # xref of the actual image
	x1 = img[1]  # xref of the mask if > 0
	b = doc.xrefStream(x0)  # read binary stream
	doc.updateStream(x0, b)  # and rewrite, compression is automatic
	if x1 > 0:  # only if image has a mask
		b=doc.xrefStream(x1)
		doc.updateStream(x1, b)

		
>>> doc.save("compressed.pdf")  # this works
>>> doc.save("compressed2.pdf", deflate=True)  # this one two!
```

**Comment by JorjMcKie** [comment]
Just got response from MuPDF:
We were not the first ones to report this. There already is a fix sitting in their development version.
This will presumably roll out with the next MuPDF version.

**Comment by iWHOswald** [comment]
Sounds fantastic. Thanks for the help! Pymupdf has been extremely useful for some projects I've been working on.

**Comment by JorjMcKie** [comment]
I just generated my local MuPDF version with the change mentioned in my previous post and comple PyMuPDF on top.
Everything worked fine!

**Comment by iWHOswald** [comment]
This is not too urgent as v 1.16 works okay for what I need to do. But thanks again for helping diagnose and fixing this issue!

**Comment by JorjMcKie** [comment]
Resolved by v1.17.4, which was just uploaded to PyPI.
